### ISSUE TITLE
Primavera CustomField IDs

### ISSUE THREAD

**Issue by darthWes** [joniles/mpxj#291]
I'm not sure who wants this, but I ended up modifying my mpxj recently to grab the id from the custom fields in my PrimaVera projects.

Is this something that you would like to add to MPXJ proper?

Coincidentally, can someone ELI5 why I can't attach my .diff/.patch files?  I'm trying to make this as simple as possible for Jon.  I'm using MPXJ 10.04 as the base for this changeset.

*************************

(net.sf.mpxj -> CustomField)

102a103,119
>      * WSL: We need the fieldId so we can
>      * clearly identify the custom field used in P6.
>      */
>     public int getFieldId()
>     {
>         return m_field_id;
>     }
> 
>     public void setFieldId(int fieldId)
>     {
>         m_field_id = fieldId;
>     }
> 
>     /**
>      * WSl, end customization
>      */
>     /**
122a140
>     private int m_field_id; //WSL: Added fieldId

*************************

(net.sf.mpxj.primavera -> PrimaVeraReader)

288c288
<             addUserDefinedField(fieldType, UserFieldDataType.valueOf(fieldDataType), fieldName);
---
>                 addUserDefinedField(fieldType, UserFieldDataType.valueOf(fieldDataType), fieldName, fieldId); //WSL: Added fieldId
1044c1044,1045
<    private void addUserDefinedField(FieldTypeClass fieldType, UserFieldDataType dataType, String name)
---
>     //WSL: Added fieldId here:
>     private void addUserDefinedField(FieldTypeClass fieldType, UserFieldDataType dataType, String name, int fieldId)
1059,1060c1060,1061
<                m_project.getCustomFields().getCustomField(taskField).setAlias(name);
< 
---
>                     //WSL: Set the fieldId
>                     m_project.getCustomFields().getCustomField(taskField).setAlias(name).setFieldId(fieldId);
1071,1072c1072,1073
<                m_project.getCustomFields().getCustomField(resourceField).setAlias(name);
< 
---
>                     //WSL: Set the fieldId
>                     m_project.getCustomFields().getCustomField(resourceField).setAlias(name).setFieldId(fieldId);
1083,1084c1084,1085
<                m_project.getCustomFields().getCustomField(assignmentField).setAlias(name);
< 
---
>                     //WSL: Set the fieldId
>                     m_project.getCustomFields().getCustomField(assignmentField).setAlias(name).setFieldId(fieldId);

For ease of use, here's the zipped patch files:
[Customizations.zip](https://github.com/joniles/mpxj/files/8772749/Customizations.zip)

**Comment by joniles** [comment]
Hello! Many thanks for the diff, that's a great idea. I struggled to get the diffs working directly, but I ended up refactoring the code in this area generally and adding support for reading/writing the Unique ID attribute for PMXML files. I've added a note to the changelog to indicate that this work is based on your suggested changes. Hope that's ok! (You can see the changes here https://github.com/joniles/mpxj/pull/292)
